File: Java/baekjoon/화목한스터디/src/스택/_1725_히스토그램.java
//baekjoon source = "https://www.acmicpc.net/problem/1725"
package 스택;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Stack;
import java.util.StringTokenizer;

public class _1725_히스토그램 {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

		StringTokenizer st;
		int n = Integer.parseInt(br.readLine());
		Stack<Integer> stack = new Stack<>();
		int max = Integer.MIN_VALUE;
		int check = 0;
		
		
		for (int i = 0; i < n; ++i) {
			st = new StringTokenizer(br.readLine());
			int number = Integer.parseInt(st.nextToken());
			check = Math.max(check, number);
			if (stack.isEmpty()) {
				stack.add(number);
			}else {
				
			}
		}
		
		bw.flush();
	}
}

File: Python/화목한스터디/1725_히스토그램.py
# baekjoon source = "https://www.acmicpc.net/problem/1725"
import itertools





n = int(input())
stack = []
numbers = [0] * n
for i in range(n):
    numbers[i] = int(input())
result = float('-inf')
i = 0
while i < n:
    while stack and numbers[stack[len(stack) - 1]] > numbers[i]:
        p = stack.pop()
        h = numbers[p]
        w = i
        if stack:
            w = i - stack[len(stack) - 1] - 1
        result = max(result, h * w)

    stack.append(i)
    i += 1
while stack:
    p = stack.pop()
    h = numbers[p]
    w = n
    if stack:
        w = i - stack[len(stack) - 1] - 1
    result = max(result, h * w)
print(result)
